homeassistant/components/tts_homely/tts.py

```diff
@@ -0,0 +1,199 @@
+    vol.Optional(CONF_SPEED, ATTR_CREDENTIALS): cv.string,
+    vol.Optional(CONF_VOICE_TYPE, ATTR_CREDENTIALS): cv.string,
+})
+
+
+
+async def async_get_engine(hass, config):
+    """Set up Homely speech component."""    
+    return HomelyProvider(hass, config)
+
+
+class HomelyProvider(Provider):
+    """The Homely speech API provider."""
+
+    def __init__(self, hass, config):
+        """Init Homely TTS service."""                 
+        self.config = config
+        if self.config.get(CONF_API_KEY): 
+            self.api_key= self.config.get(CONF_API_KEY) 
+        else: 
+            self.api_key=''
+        
+        if self.config.get(CONF_SPEED):
+            self.speed = self.config.get(CONF_SPEED)
+        else:
+            self.speed = DEFAULT_SPEED
+            
+        
+        if self.config.get(CONF_VOICE_TYPE):
+            self.voicetype = voice_list.get(self.config.get(CONF_VOICE_TYPE))
+        else:
+            self.voicetype = DEFAULT_VOICE
+        self.hass = hass
+        self._lang = self.config.get(CONF_LANG)
+        self.headers = {
+            REFERER: "http://translate.google.com/",
+            USER_AGENT: ("Mozilla/5.0 (Windows NT 10.0; WOW64) "
+                         "AppleWebKit/537.36 (KHTML, like Gecko) "
+                         "Chrome/47.0.2526.106 Safari/537.36"),
+            'api_key':self.api_key,
+            'speed':self.speed,
+            'prosody':'1',
+            'voice': self.voicetype
+        }
+        _LOGGER.error(self.api_key)
+        self.name = 'Homely'
+        
+
+    @property
+    def default_language(self):
+        """Return the default language."""
+        return self._lang
+
+    @property
+    def supported_languages(self):
+        """Return list of supported languages."""
+        return SUPPORT_LANGUAGES
+
+    async def async_get_tts_audio(self, message, language, options=None):
+        """Load TTS from google or FPT."""
+        from gtts_token import gtts_token
+
+        token = gtts_token.Token()
+        websession = async_get_clientsession(self.hass)
+        message_parts = self._split_message_to_parts(message)
+
+        data = b''
+        for idx, part in enumerate(message_parts):
+            part_token = await self.hass.async_add_job(
+                token.calculate_token, part)
+            
+            url_param = {
+                'ie': 'UTF-8',
+                'tl': language,
+                'q': yarl.URL(part).raw_path,
+                'tk': part_token,
+                'total': len(message_parts),                
+                'idx': idx,
+                'client': 'tw-ob',
+                'textlen': len(part),
+            }
+            
+
+            try:
+                with async_timeout.timeout(10):
+                    
+                    if self._lang =="vi" and self.api_key!='':
+                        url_mp3 = requests.post(FPT_SPEECH_URL, data = part.encode('utf-8'), headers = self.headers).json()['async']
```

Comment:
Please extract the service specific interface code and data parsing to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_component_code_review.html#4-communication-with-devices-services